manager/controlapi/node.go

```diff
@@ -220,14 +240,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 	}
 
 	if demote && s.raft != nil {
-		memberlist := s.raft.GetMemberlist()
-		for raftID, member := range memberlist {
-			if member.NodeID == request.NodeID {
-				if err := s.raft.RemoveMember(ctx, raftID); err != nil {
-					return nil, err
+		if err := s.raft.RemoveMember(ctx, member.RaftID); err != nil {
+			// Rollback to the initial Spec if we can't finalize the role
+			// change by removing the member from raft.
+			s.store.Update(func(tx store.Tx) error {
+				node = store.GetNode(tx, request.NodeID)
+				if node == nil {
+					return nil
 				}
-				break
-			}
+
+				node.Spec = initSpec
```

Comment:
It looks like this is a rollback that happens if we tried to remove the manager from the raft cluster, but failed. I think you're right that this rollback won't happen if the leader dies between the two transactions.

I'm thinking maybe we shouldn't do this rollback, for a few reasons:

- The client may have changed other things in the spec. Here we're discarding all those other changes because the demotion failed.
- It's a spec, so it should be under control of the user. We shouldn't be changing things based on actual state.
- The corner case @stevvooe points out here with node failure.